### PR TITLE
Gh 57 incorrect ns3004 for derived types

### LINKED ISSUE #57: Incorrect NS3004 for derived types
First of all, the analyzer is really useful - thanks for creating the project!

I think I'm incorrectly getting the `Could not set value of type … to argument ... because the types are incompatible` error when trying to assign a `Dictionary` to a `IReadOnlyDictionary` argument and was able to reproduce it with this code:
```C#
using NSubstitute;
using Xunit;

public class ExampleTest
{
    public interface IExampleInterface
    {
        bool Method(out object value);
    }

    [Fact]
    public void DictionaryTest()
    {
        IExampleInterface substitute = Substitute.For<IExampleInterface>();
        substitute.Method(out Arg.Any<object>())
            .Returns(ci =>
            {
                ci[0] = "string";
                return true;
            });

        bool result = substitute.Method(out object value);

        Assert.True(result);
        Assert.Equal("string", value);
    }
}
```

The line with `ci[0] = "string";` triggers the analyser error, however, the test succeeds in passing (i.e. NSubstitute does correctly set the parameter).

I think the problem is we're checking if the type is equal instead of if the assignment will work in [AbstractCallInfoAnalyzer.cs](https://github.com/nsubstitute/NSubstitute.Analyzers/blob/632c99f9e6156ac3dd264d054aa4c039bdaff218/src/NSubstitute.Analyzers.Shared/DiagnosticAnalyzers/AbstractCallInfoAnalyzer.cs#L248) - maybe `Equals` should be replaced with a call to [CanCast](https://github.com/nsubstitute/NSubstitute.Analyzers/blob/632c99f9e6156ac3dd264d054aa4c039bdaff218/src/NSubstitute.Analyzers.Shared/DiagnosticAnalyzers/AbstractCallInfoAnalyzer.cs#L66)?

### PR DESCRIPTION
**Pull request by tpodolak**
Apologies for the size of the PR, however, I also had to fix `callInfo.ArgAt<T>` and `callInfo.Arg<T>` analysis which had the same bug 
Closes #57

### CODE REVIEW

File: src/NSubstitute.Analyzers.Shared/DiagnosticAnalyzers/AbstractCallInfoAnalyzer.cs [hunk 1]
```diff
@@ -245,7 +247,7 @@ private bool AnalyzeAssignment(SyntaxNodeAnalysisContext syntaxNodeContext, ILis
                 }
 
                 var typeInfo = syntaxNodeContext.SemanticModel.GetTypeInfo(assignmentExpressionSyntax);
-                if (typeInfo.Type != null && Equals(typeInfo.Type, substituteCallParameters[position.Value].Type) == false)
+                if (typeInfo.Type != null && IsAssignableTo(syntaxNodeContext.Compilation, typeInfo.Type, substituteCallParameters[position.Value].Type) == false)
```
Comment (dtchepak) [inline src/NSubstitute.Analyzers.Shared/DiagnosticAnalyzers/AbstractCallInfoAnalyzer.cs]: Is there a reason for `IsAssignableTo(...) == false` vs `!IsAssignableTo(..)`?

File: src/NSubstitute.Analyzers.CSharp/DiagnosticAnalyzers/CallInfoAnalyzer.cs [hunk 2]
```diff
@@ -86,5 +86,11 @@ protected override bool CanCast(Compilation compilation, ITypeSymbol sourceSymbo
         {
             return compilation.ClassifyConversion(sourceSymbol, destinationSymbol).Exists;
         }
+
+        protected override bool IsAssignableTo(Compilation compilation, ITypeSymbol fromSymbol, ITypeSymbol toSymbol)
+        {
+            var conversion = compilation.ClassifyConversion(fromSymbol, toSymbol);
+            return conversion.Exists && conversion.IsExplicit == false && conversion.IsNumeric == false;
```
Comment (dtchepak) [inline src/NSubstitute.Analyzers.CSharp/DiagnosticAnalyzers/CallInfoAnalyzer.cs]: Can you please explain why `conversion.IsNumeric` is excluded? (I don't know anything about `Conversion` so sorry if this is a silly question.)

As per other comment, is there a reason for preferring `x == false` compared with `!x`? (I don't have a strong view on this, just curious)

Comment (tpodolak) [reply src/NSubstitute.Analyzers.CSharp/DiagnosticAnalyzers/CallInfoAnalyzer.cs]: There is implicit conversion between some numeric types. For instance this is valid c#
````c#
int x = 1;
decimal xx = x;
````
so conversion is numeric and exists. However it doesnt mean that type is assignable to other type in terms of NSubstitute check - as NSubstitute checks this as `type.IsAssignableFrom(otherType)`

Here is a simple repro
````c#
using NSubstitute;
using System.Collections.Generic;
using NSubstitute.ReturnsExtensions;
using Xunit;

namespace MyNamespace
{
    public interface IFoo
    {
        int Bar(out int x);
    }

    public class FooTests
    {
        [Fact]
        public void Test()
        {
            var substitute = Substitute.For<IFoo>();

            int i;
            substitute.Bar(out i).Returns(callInfo =>
            {
                callInfo.ArgAt<decimal>(0);
                return 1;
            });

            substitute.Bar(out var x);
        }
    }
}
````
rider64_2019-01-02_21-54-24

File: src/NSubstitute.Analyzers.Shared/DiagnosticAnalyzers/AbstractCallInfoAnalyzer.cs (hunk 1 above)
Comment (tpodolak) [reply src/NSubstitute.Analyzers.Shared/DiagnosticAnalyzers/AbstractCallInfoAnalyzer.cs]: This is just a personal preference. I find it easier to spot compared to ! sign.

**Comment by dtchepak** [comment]
Thanks @tpodolak. I've merged this. 

I disabled the code coverage check requirement as the status reporting does not seem to be working correctly. I think it's ok to manually review the coverage information reported in the PR and we can raise/resolve any issues with it as part of the review process. Sound ok?

**Pull request merged**